src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c

```diff
@@ -0,0 +1,1042 @@
+#define XOCL_XGQ_CONFIG_TIME 	msecs_to_jiffies(10 * 1000) 
+#define XOCL_XGQ_MSLEEP_1S	(1000)      //1 s
+
+/*
+ * Note:
+ * We should keep the following flags in-sync with RPU device
+ * so that flags passed with XGQ commands(loadxclbin or configure)
+ * will be able to be recongized by RPU.
+ */
+typedef enum xrt_xgq_pkt_type {
+	XRT_XGQ_PKT_TYPE_UNKNOWN = 0,
+	XRT_XGQ_PKT_TYPE_PDI,
+	XRT_XGQ_PKT_TYPE_XCLBIN,
+	XRT_XGQ_PKT_TYPE_AF,
+} xrt_xgq_pkt_type_t;
+
+typedef union {
+	struct {
+		u32 version:16;
+		u32 type:16;
+		u32 cid:16;
+		u32 rcode:16;
+	};
+	u32 head[2];
+} xgq_vmr_head_t;
+
+struct xgq_vmr_payload_xclbin {
+	u32 address;
+	u32 size;
+};
+
+struct xgq_vmr_payload_af {
+	u32 buffer_addr;
+	u32 buffer_size;
+};
+
+struct xgq_vmr_pkt {
+	xgq_vmr_head_t head;
+	union {
+		struct xgq_vmr_payload_xclbin payload_xclbin;
+		struct xgq_vmr_payload_af payload_af;
+	};
+};
+
+typedef void (*xocl_xgq_complete_cb)(void *arg, struct xrt_com_queue_entry *ccmd);
+
+struct xocl_sq {
+	struct xrt_sub_queue_entry 	xgq_sq_head;
+	struct xgq_vmr_pkt 		xgq_sq_pkt;
+};
+
+struct xocl_xgq_cmd {
+	struct xocl_sq		xgq_sq;
+	struct list_head	xgq_cmd_head;
+	struct completion	xgq_cmd_complete;
+	xocl_xgq_complete_cb    xgq_cmd_cb;
+	void			*xgq_cmd_arg;
+	int			xgq_cmd_rcode;
+	struct timer_list	xgq_cmd_timer;
+	struct xocl_xgq		*xgq;
+	u64			xgq_cmd_timeout_jiffies; /* timout till */
+};
+
+struct xocl_xgq;
+
+struct xgq_worker {
+	struct task_struct	*complete_thread;
+	bool			error;
+	bool			stop;
+	struct xocl_xgq		*xgq;
+};
+
+struct xocl_xgq {
+	struct platform_device 	*xgq_pdev;
+	struct xgq	 	xgq_queue;
+	u64			xgq_io_hdl;
+	void __iomem		*xgq_ring_base;
+	u32			xgq_slot_size;
+	void __iomem		*xgq_sq_base;
+	void __iomem		*xgq_cq_base;
+	struct mutex 		xgq_lock;
+	int 			xgq_polling;
+	u32			xgq_intr_base;
+	u32			xgq_intr_num;
+	struct list_head	xgq_submitted_cmds;
+	struct completion 	xgq_irq_complete;
+	struct xgq_worker	xgq_complete_worker;
+	struct xgq_worker	xgq_health_worker;
+	bool			xgq_halted;
+	bool			xgq_ospi_inuse;
+	int 			xgq_cmd_id;
+};
+
+/*
+ * when detect cmd is completed, find xgq_cmd from submitted_cmds list
+ * and find cmd by cid; perform callback and remove from submitted_cmds.
+ */
+static void cmd_complete(struct xocl_xgq *xgq, struct xrt_com_queue_entry *ccmd)
+{
+	struct xocl_xgq_cmd *xgq_cmd;
+	struct list_head *pos, *next;
+
+	list_for_each_safe(pos, next, &xgq->xgq_submitted_cmds) {
+		xgq_cmd = list_entry(pos, struct xocl_xgq_cmd, xgq_cmd_head);
+
+		if (xgq_cmd->xgq_sq.xgq_sq_head.cid == ccmd->cid) {
+			XGQ_INFO(xgq, "xgq cmd %d completed", ccmd->cid);
+
+			list_del(pos);
+
+			if (xgq_cmd->xgq_cmd_cb)
+				xgq_cmd->xgq_cmd_cb(xgq_cmd->xgq_cmd_arg, ccmd);
+			return;
+		}
+
+	}
+
+	XGQ_WARN(xgq, "unknown cid %d received", ccmd->cid);
+	return;
+}
+
+/*
+ * Read completed cmd based on XGQ protocol.
+ */
+void read_completion(struct xrt_com_queue_entry *ccmd, u64 addr)
+{
+	u32 i;
```

Comment:
Uninitialized variable
```suggestion
	u32 i = 0;
```